### ISSUE TITLE
Do data transforms happen on every yield from the train loader or once at load time?

### ISSUE THREAD

**Issue by catskillsresearch** [Project-MONAI/tutorials#51]
**QUESTION 1:**

 When I apply a list of transforms as in the Spleen tutorial notebook, do they happen once here:

`train_ds = CacheDataset(data=train_files, transform=train_trans, cache_rate=1.0, num_workers=8)`

Note that it says

`Load and cache transformed data: 100%|██████████| 41/41 [00:15<00:00,  2.65it/s]`

The past tense "transformed" seems to indicate that transformations only happen once.  Or, after defining

`train_loader = DataLoader(train_ds, batch_size=2, shuffle=True, num_workers=loader_workers)`

do the transformations actually happen on every reference to an item in the training queue, specifically here:

`        for batch_data in train_loader:`

This is the ideal case for me.  In the former case, should I repeat my data 100 times before running it through `CacheDataset` to get my augmentations?  Is that standard?  It seems it would be a lot better to do the transformations on the fly.  Also very necessary for a subsampling transformation like `RandCropByPosNegLabeld`.

This could be a dumb question, I just don't see it spelled out in the docs and the logging printed out by `CacheDataset`.

**NOTE:** I'm guessing this happens with every `train_loader` yield, because my training loop has slowed way down.  This leads to

**QUESTION 2:** Would it be possible to do these transforms in the GPU?  I'm assuming the slowdown happens because they are on CPU, as shown by the attached picture, which depicts a very lightly loaded GPU and 1 hammered CPU core.   This leads to 

![train_loader](https://user-images.githubusercontent.com/38082822/97381377-59c6ce80-189f-11eb-9603-9ad840942fc0.png)

**QUESTION 3:** Can I speed up the train loader transformations by adding workers?  I'm guessing Yes.  If not, should be Yes.  I'll try it now.

**ANSWER 1&3:** Yes it must be happening for each `train_loader` yield, yes adding workers helps.  NOTE: A comment in the tutorial notebook says "because this is cached in memory, you only need one work".   This is misleading.  And on Question 2: The 8 cores I added are 100% active.  The GPU is  10% to 25% loaded max.  **These transforms should happen on the GPU!!**  Most of the compute time is spent in the transforms.  Very little in the training.

**Comment by ericspod** [comment]
What `CacheDataset` will do is cache the results from the transform sequence up to the first transform which is randomized or is not recognized as deterministic. For question 1 this means part of the transform sequence is run on each data request. For question 2 we're exploring how to do transforms in GPU, alternatively `ThreadBuffer` can be used to parallelize the batch generation with training which can speed training up if the times for these two operations are similar. I have a PR for a demonstration notebook on this repo now. For question 3 adding workers will speed things up, though probably only up to however many physical cores you have.

**Comment by catskillsresearch** [comment]
When the DRAM goes up as in the 3090 it will be possible to do a lot more on the GPU, so more transforms should be pipelined into the GPU.  I guess I'm asking that whether NVidia programs a library with transforms they should always make a GPU-enabled version, in anticipation for when those large DRAM boards become more prevalent. 

As a case in point, someone actually ran a program entirely in DRAM: https://www.techradar.com/news/nvidia-rtx-3090-is-such-a-monster-it-can-have-crysis-3-installed-in-vram-and-run-it-smoothly-at-4k

**Comment by Nic-Ma** [comment]
Hi @catskillsresearch ,

Thanks for your suggestion.
Some of the transforms can run on GPU now, most of them are still based on numpy, so they can only run on CPU.
We will try to enable more transforms on GPU later when **PyTorch tensor operations** become more stable and faster.

Thanks.
